common/client/node_lifecycle.go

```diff
@@ -322,46 +413,114 @@ func (n *node[CHAIN_ID, HEAD, RPC]) outOfSyncLoop(isOutOfSync func(num int64, td
+
+			if noNewHeadsTimeoutThreshold > 0 {
+				headsSub.ResetTimer(noNewHeadsTimeoutThreshold)
 			}
-			lggr.Debugw(msgReceivedBlock, "blockNumber", head.BlockNumber(), "blockDifficulty", head.BlockDifficulty(), "nodeState", n.getCachedState())
-		case <-time.After(zombieNodeCheckInterval(n.chainCfg.NodeNoNewHeadsThreshold())):
+
+			lggr.Debugw(msgReceivedBlock, "blockNumber", head.BlockNumber(), "blockDifficulty", head.BlockDifficulty(), "syncIssues", syncIssues)
+		case <-time.After(zombieNodeCheckInterval(noNewHeadsTimeoutThreshold)):
 			if n.poolInfoProvider != nil {
 				if l, _ := n.poolInfoProvider.LatestChainInfo(); l < 1 {
-					lggr.Critical("RPC endpoint is still out of sync, but there are no other available nodes. This RPC node will be forcibly moved back into the live pool in a degraded state")
+					lggr.Critical("RPC endpoint is still out of sync, but there are no other available nodes. This RPC node will be forcibly moved back into the live pool in a degraded state", "syncIssues", syncIssues)
 					n.declareInSync()
 					return
 				}
 			}
-		case err := <-sub.Err():
-			lggr.Errorw("Subscription was terminated", "nodeState", n.getCachedState(), "err", err)
+		case err := <-headsSub.Errors:
+			lggr.Errorw("Subscription was terminated", "err", err)
+			n.declareUnreachable()
+			return
+		case <-headsSub.NoNewHeads:
+			// we are not resetting the timer, as there is no need to add syncIssueHeadIsNotIncreasing until it's removed on new head.
+			syncIssues |= syncIssueHeadIsNotIncreasing
+			lggr.Debugw(fmt.Sprintf("No new heads received for %s. Node stays out-of-sync due to sync issues: %s", noNewHeadsTimeoutThreshold, syncIssues))
+		case latestFinalized, open := <-finalizedHeadsSub.Heads:
+			if !open {
+				lggr.Errorw("Finalized heads subscription channel unexpectedly closed")
+				n.declareUnreachable()
+				return
+			}
+			if !latestFinalized.IsValid() {
+				lggr.Warn("Latest finalized block is not valid")
+				continue
+			}
+
+			receivedNewHead := n.onNewFinalizedHead(lggr, &localHighestChainInfo, latestFinalized)
+			if !receivedNewHead {
+				continue
+			}
+
+			syncIssues &= ^syncIssueFinalizedHeadIsNotIncreasing
+			if noNewFinalizedBlocksTimeoutThreshold > 0 {
+				finalizedHeadsSub.ResetTimer(noNewFinalizedBlocksTimeoutThreshold)
+			}
+
+			lggr.Debugw(msgReceivedFinalizedBlock, "blockNumber", latestFinalized.BlockNumber(), "syncIssues", syncIssues)
+		case err := <-finalizedHeadsSub.Errors:
+			lggr.Errorw("Finalized head subscription was terminated", "err", err)
 			n.declareUnreachable()
 			return
+		case <-finalizedHeadsSub.NoNewHeads:
+			// we are not resetting the timer, as there is no need to add syncIssueFinalizedHeadIsNotIncreasing until it's removed on new finalized head.
+			syncIssues |= syncIssueFinalizedHeadIsNotIncreasing
+			lggr.Debugw(fmt.Sprintf("No new finalized heads received for %s. Node stays out-of-sync due to sync issues: %s", noNewFinalizedBlocksTimeoutThreshold, syncIssues))
```

Comment:
nit: this may not matter but shouldn't this be %d
```suggestion
			lggr.Debugw(fmt.Sprintf("No new finalized heads received for %d. Node stays out-of-sync due to sync issues: %s", noNewFinalizedBlocksTimeoutThreshold, syncIssues))
```

Reply:
With `%d` it will be formatted as int, so `10s` will be printed as `100000000`

Reply:
Oh right sorry forgot this was `time.Duration`